File: 03-serialization/json/java/MyApp.java
import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;

import java.util.Arrays;

public class MyApp {
    private static final String[] CAMPOS = {"Name", "EnrollNumber", "Height", "LuckNumbers"};

    // UPPER_CAMEL_CASE: name -> Name, enrollNumber -> EnrollNumber, ...
    private static final Gson GSON = new GsonBuilder()
            .setFieldNamingPolicy(FieldNamingPolicy.UPPER_CAMEL_CASE)
            .create();

    public static void main(String[] args) {
        if (args.length > 0) {
            receber(args[0]);
        } else {
            enviar();
        }
    }

    private static void enviar() {
        Person p = new Person();
        p.setName("Ada Lovelace");
        p.setEnrollNumber(100001);
        p.setHeight(1.68f);
        p.setLuckNumbers(Arrays.asList(2, 11, 29));
        System.out.println(GSON.toJson(p));
    }

    private static void receber(String json) {
        JsonObject obj;
        try {
            obj = JsonParser.parseString(json).getAsJsonObject();
        } catch (JsonSyntaxException | IllegalStateException ex) {
            System.out.println("JSON inválido: " + ex.getMessage());
            System.exit(1);
            return;
        }

        for (String campo : CAMPOS) {
            if (!obj.has(campo)) {
                System.out.printf("JSON incompleto: falta a chave \"%s\"\n", campo);
                System.exit(1);
            }
        }

        Person p = GSON.fromJson(obj, Person.class);
        System.out.printf("Name: %s\n", p.getName());
        System.out.printf("EnrollNumber: %d\n", p.getEnrollNumber());
        System.out.printf("Height: %.2f\n", p.getHeight());
        System.out.printf("LuckNumbers: %s\n", p.getLuckNumbers());
    }
}

File: 03-serialization/json/python/MyApp.py
from __future__ import annotations  # list[int] em anotações no Python 3.8

from dataclasses import asdict, dataclass, field
import json
import sys

CAMPOS = ("Name", "EnrollNumber", "Height", "LuckNumbers")


@dataclass
class Person:
    Name: str
    EnrollNumber: int
    Height: float
    LuckNumbers: list[int] = field(default_factory=list)


def enviar():
    p = Person(
        Name="Grace Hopper",
        EnrollNumber=200002,
        Height=1.60,
        LuckNumbers=[5, 13, 41],
    )
    print(json.dumps(asdict(p), separators=(",", ":"), ensure_ascii=False))


def receber(texto):
    try:
        dados = json.loads(texto)
    except json.JSONDecodeError as ex:
        sys.exit(f"JSON inválido: {ex}")

    if not isinstance(dados, dict):
        sys.exit(f"JSON inválido: esperava um objeto, veio {type(dados).__name__}")

    faltando = [c for c in CAMPOS if c not in dados]
    if faltando:
        sys.exit(f"JSON incompleto: falta(m) a(s) chave(s) {', '.join(faltando)}")

    p = Person(**{c: dados[c] for c in CAMPOS})
    print(f"Name: {p.Name}")
    print(f"EnrollNumber: {p.EnrollNumber}")
    print(f"Height: {p.Height}")
    print(f"LuckNumbers: {p.LuckNumbers}")


if __name__ == "__main__":
    if len(sys.argv) > 2:
        sys.exit("uso: python3 MyApp.py [json]")

    if len(sys.argv) == 2:
        receber(sys.argv[1])
    else:
        enviar()
